torch/_higher_order_ops/associative_scan.py

```diff
@@ -71,49 +166,110 @@ def add(x: torch.Tensor, y: torch.Tensor):
+    check_args(combine_fn, leaves, spec, dim)
 
-    for x in leaves[1:]:
-        assert x.shape == shape, "All input tensors must have the same shape"
+    if reverse:
+        leaves = [torch.flip(elem, [dim]) for elem in leaves]
 
     combine_fn = functools.partial(
         wrap_combine_fn_flat, combine_fn=combine_fn, spec=spec, num_leaves=len(leaves)
     )
 
-    result_flat = associative_scan_op(combine_fn, leaves, dim)
+    if generic_scan:
+        result_flat = generic_associative_scan(combine_fn, leaves, dim, lifted_args)
+    else:
+        result_flat = associative_scan_op(combine_fn, leaves, dim, lifted_args)
+
+    if reverse:
+        result_flat = [torch.flip(elem, [dim]) for elem in result_flat]
 
     return pytree.tree_unflatten(result_flat, spec)
 
 
+def generic_associative_scan(operator, elems_flat, dim=0, lifted_args=()):
+    # TODO: The recursion involved here "unrolls" the scan
+    # function for all inputs. Could there be a more efficient
+    # way instead of running over the operation in sequence?
+    def _scan(elems):
+        """Perform scan on `elems`."""
+        num_elems = elems[0].shape[dim]
+
+        if num_elems < 2:
+            return elems
+
+        reduced_elems = operator(
+            *[elem[slice_along_axis(0, -1, stride=2, dim=dim)] for elem in elems],
+            *[elem[slice_along_axis(1, None, stride=2, dim=dim)] for elem in elems],
+            *lifted_args,
+        )
+
+        # Recursively compute scan for partially reduced tensors.
+        odd_elems = _scan(reduced_elems)
+
+        if num_elems % 2 == 0:
+            even_elems = operator(
+                *[e[slice_along_axis(0, -1, dim=dim)] for e in odd_elems],
+                *[e[slice_along_axis(2, None, stride=2, dim=dim)] for e in elems],
+                *lifted_args,
+            )
+        else:
+            even_elems = operator(
+                *odd_elems,
+                *[e[slice_along_axis(2, None, stride=2, dim=dim)] for e in elems],
+                *lifted_args,
+            )
+
+        # The first element of a scan is the same as the first element
+        # of the original `elems`.
+        even_elems = [
+            torch.cat([elem[slice_along_axis(0, 1, dim=dim)], result], dim=dim)
+            if result.shape.numel() > 0 and elem.shape[dim] > 0
+            else result
+            if result.shape.numel() > 0
+            else elem[
+                slice_along_axis(0, 1, dim=dim)
+            ]  # Jax allows/ignores concat with 0-dim, Pytorch does not
+            for (elem, result) in zip(elems, even_elems)
+        ]
+
+        return list(
+            safe_map(functools.partial(_interleave, dim=dim), even_elems, odd_elems)
+        )
+
+    scans = _scan(elems_flat)
+
+    return scans
+
+
 associative_scan_op = HigherOrderOperator("associative_scan")
 
 
 def trace_associative_scan(
-    proxy_mode, func_overload, combine_fn: Callable, input: List[torch.Tensor], dim: int
+    proxy_mode,
```

Comment:
I believe lifted_args should have its separate pr. We should only keep changes related to supporting "generic_associtiave_scan" in this PR. Otherwise it would be difficult to review and track.